mysys/my_uuid_v7.c

```diff
@@ -0,0 +1,139 @@
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1335  USA */
+
+/*
+  implements Universal Unique Identifiers version 7, as described in
+  draft-ietf-uuidrev-rfc4122bis-14.
+
+  A UUIDv7 has the following structure:
+
+  Field                     Octet #          Note
+ unix_ts_ms                 0-5     Big-endian unsigned number of
+                                    Unix epoch timestamp in
+                                    milliseconds.
+ version_and_rand_a         6-7     CSPRNG 12 bits multiplexed
+                                    with the version number.
+ variant_and_rand_b         8-15    CSPRNG 62 bits multiplexed
+                                    with the version number.
+*/
+
+#include "mysys_priv.h"
+#include <my_rnd.h>
+#include <m_string.h>
+#include <myisampack.h> /* mi_int2store, mi_int4store */
+
+static my_bool my_uuid_v7_inited= 0;
+static uint64 milliseconds;
+static uint64 uuid_time= 0;
+
+static mysql_mutex_t LOCK_uuid_v7_generator;
+
+#define UUID_VERSION          0x7000
+#define UUID_VERSION_MASK     0x0FFF
+#define UUID_VARIANT          0x8000000000000000
+#define UUID_VARIANT_MASK     0x3FFFFFFFFFFFFFFF
+
+/**
+  Init structures needed for my_uuid_v7
+*/
+void my_uuid_v7_init()
+{
+  if (my_uuid_v7_inited)
+    return;
+  my_uuid_v7_inited= 1;
+  milliseconds= 0;
+
+  mysql_mutex_init(key_LOCK_uuid_v7_generator, &LOCK_uuid_v7_generator, MY_MUTEX_INIT_FAST);
+}
+
+/**
+   Create a global unique identifier version 7(uuidv7)
+
+   @func  my_uuid_v7()
+   @param to   Store uuidv7 here. Must be of size MY_UUID_SIZE (16)
+*/
+void my_uuid_v7(uchar *to)
+{
+  uint64 tv, unix_ts_ms;
+  uint16 rand_a, version_and_rand_a;
+  uint64 rand_b, variant_and_rand_b;
+
+  DBUG_ASSERT(my_uuid_v7_inited);
+
+  tv= my_hrtime().val / 1000;
+
+  // Regulate the access to uuid_time and milliseconds
+  mysql_mutex_lock(&LOCK_uuid_v7_generator);
+
+  if (likely(tv > uuid_time))
+  {
+    /*
+      Current time is ahead of last timestamp, as it should be.
+      If we "borrowed time", give it back, just as long as we
+      stay ahead of the previous timestamp.
+    */
+    if (milliseconds)
+    {
+      uint64 delta;
+      /*
+        -1 so we won't make tv= uuid_time for milliseconds >= (tv - uuid_time)
+      */
+      delta= MY_MIN(milliseconds, (uint16)(tv - uuid_time -1));
+      tv-= delta;
+      milliseconds-= delta;
+    }
+  }
+  else if (unlikely(tv <= uuid_time))
+  {
+    /*
+      If several requests for UUIDs end up on the same tick
+      or if the system clock is turned *back*.
+      We add milli-seconds to make them different.
+      ( current_timestamp + milliseconds * calls_in_this_period )
+      may end up > next_timestamp; this is OK. Nonetheless, we'll
+      try to unwind milliseconds when we get a chance to.
+    */
+    milliseconds += uuid_time - tv + 1;
+    tv = uuid_time + 1;
+  }
```

Comment:
If you agree that https://github.com/MariaDB/server/pull/3093/files/46bef2ac01750635ea81a733d7664c853154c8ae#r1515136655 is a concern, I think it could be fixed here with…

```c
if (milliseconds > 1000)
{
   /* We are building up too much "borrowed time", >1000 milliseconds.
    * The UUID7() output could become non-time-sortable if the server
    * process restarts, causing `milliseconds` to reset to 0.
    *
    * Sleep until we have less than HALF of that maximum debt (500
    * milliseconds), in order to avoid repeatedly blocking and sleeping
    * on successive calls.
    */
    my_sleep((milliseconds - 500) * 1000UL);
    … recalculate milliseconds, TV …
}
```

Reply:
I agree that restarting the server after having built a lot of borrowed time could break time-sortability.

A solution that I would like more would be implementing the [Replace Left-Most Random Bits with Increased Clock Precision](https://datatracker.ietf.org/doc/draft-ietf-uuidrev-rfc4122bis/) variation from the UUID specs.

Encoding 10 more bits of timestamp would allow us to have a resolution of microseconds instead of milliseconds. 

With the current implementation, generating 1 million UUIDv7 results in 1000 seconds of borrowed time. With the additional time precision, instead of borrowing milliseconds, we would borrow microseconds. This means that generating 1 million UUIDv7 would only borrow 1 second.
Even with my solution breaking time sortability would still be possible but it would require the database to shut down right after generating a humungous amount of UUIDv7s.

What I don't like about your solution is that it puts an upper bound to the number of UUIDv7 that can be generated in 1 second and that upper bound might be too low for some use cases. How common is it to generate more than 1000 UUIDs per second on average? If I had to speculate I would say that it is not uncommon.

What do you think? @dlenski